docs/installing/_index.md: quickstart intro (#219)
flatcar-linux/flatcar-docs: docs/installing/_index.md

    - quickstart
---

This guide aims to get you up and running in a few minutes.
We'll cover
- concepts, configuration, and provisioning
- writing a basic Flatcar configuration and testing it locally with qemu
- further reading - pointers to docs for specific cloud providers

### Concepts, Configuration, and Provisioning

Flatcar Container Linux is configured at _provisioning time_.
There are two configuration languages to set up Flatcar, aimed at different use cases:
- [Container Linux Config][cl-config] (CLC) is human-readable / writable YAML and must be converted (transpiled) into Ignition config before Flatcar can use it.
- [Ignition config][ignition] is machine-readable JSON fed to Flatcar's ignition service.

>   Ignition is Flatcars "installation sevice" which configures a Flatcar instance during provisioning.

Reviewer: ```suggestion
  Ignition is Flatcars "installation service" which configures a Flatcar instance during provisioning.
```

  The config file is passed via the "custom data" or "user data" option of cloud providers, and can be supplied by various mechanisms to private cloud VMs and bare metal.
  Ignition config is rarely written by a human; it's usually generated by provisioning automation or transpiled from user-written CLC.

Use CLC to customise your Flatcar deployment, e.g. to
- add custom users and groups
- create and manage storage devices, file systems, and swap, and create custom files
- customise automatic updates and define reboot windows
- create custom network(d) configurations and systemd units
  - Want to e.g. run a custom operation or start a service each time the instance boots?

>     Define a custom systemd unit via CLC.

Reviewer: ```suggestion
    Define a custom [systemd unit via CLC][systemd].
```

To convert CLC into machine-readable Ignition config, just download the latest release of the [config transpiler](https://github.com/flatcar-linux/container-linux-config-transpiler/releases) for your platform.
Alternatively, use our up-to-date [container image](https://github.com/flatcar-linux/container-linux-config-transpiler/pkgs/container/ct): `docker run --rm -i ghcr.io/flatcar-linux/ct:latest`.

Then, after converting CLC to Ignition config, pass the Ignition config along when provisioning your instance(s).
The config is then passed via "user data" / "custom data" or similar means to the provisioning logic.

#### Writing your first config and testing it locally in a qemu VM

The way from a small [Container Linux Config (CLC) YAML][cl-configs] or [Ignition JSON][ignition] file to a local [QEMU VM][qemu-docs] on your laptop is not far.
Here we will create a systemd service that starts an NGINX container as example configuration for the VM.
This is a good starting point for you to modify the CLC YAML file (or the Ignition JSON file) and test it by provisioning a temporary QEMU VM.
This should work on most Linux systems and assumes you have an SSH key set up for ssh-agent.

> First download the Flatcar QEMU image and the helper script to start it with QEMU.

Reviewer: ```suggestion
First download the Flatcar QEMU image and the helper script to start it with QEMU but don't run it yet.
```
Following it paragraph by paragraph one might accidentally boot it already if the next paragraph is not yet read.

```shell
wget https://stable.release.flatcar-linux.net/amd64-usr/current/flatcar_production_qemu_image.sh
chmod +x flatcar_production_qemu_image.sh
wget https://stable.release.flatcar-linux.net/amd64-usr/current/flatcar_production_qemu_image.img.bz2
bunzip2 flatcar_production_qemu_image.img.bz2
```

> For testing Ignition configruations we have to make sure that we always boot an unmodified fresh image because Ignition only runs on first boot.

Reviewer: ```suggestion
For Ignition configurations to be recognized we have to make sure that we always boot an unmodified fresh image because Ignition only runs on first boot.
```
typo and more general statement

Therefore, before trying to use an Ignition config we will always discard the image modifications by using a fresh copy.

> You can already boot the image and have a look around in the OS through the QEMU VGA console - you can close the QEMU window or stop the script with `Ctrl-C`.

Reviewer: ```suggestion
You can already boot the image with `./flatcar_production_qemu.sh` and have a look around in the OS through the QEMU VGA console - you can close the QEMU window or stop the script with `Ctrl-C`.
```

  The config file is passed via the "custom data" or "user data" option of cloud providers, and can be supplied by various mechanisms to private cloud VMs and bare metal.
  Ignition config is rarely written by a human; it's usually generated by provisioning automation or transpiled from user-written CLC.

Use CLC to customise your Flatcar deployment, e.g. to
- add custom users and groups
- create and manage storage devices, file systems, and swap, and create custom files
- customise automatic updates and define reboot windows
- create custom network(d) configurations and systemd units
  - Want to e.g. run a custom operation or start a service each time the instance boots?
    Define a custom [systemd unit][systemd] via CLC.
All above tasks only take a few lines of YAML and are covered in our [CLC examples][cl-examples].
For a comprehensive discussion of all options available in CLC have a look at the [CLC specification][cl-spec].

To convert CLC into machine-readable Ignition config, just download the latest release of the [config transpiler](https://github.com/flatcar-linux/container-linux-config-transpiler/releases) for your platform.

> Alternatively, use our up-to-date [container image](https://github.com/flatcar-linux/container-linux-config-transpiler/pkgs/container/ct): `docker run --rm -i ghcr.io/flatcar-linux/ct:latest`.

Reviewer 2: and available on ArchLinux repository :D https://archlinux.org/packages/community/x86_64/ct/

curl http://localhost/
```

> If you can't or don't want to use the SSH public key from ssh-agent, you can specify another one in the CLC config by adding this section to your YAML file:

Reviewer: This sentence was there to provide an alternative if the autodetection of the helper script does not work or if this guide is used in another setting. I think it's worth making it clear that above we didn't need to specify it but this is how to do it as an alternative.

Author: Gotcha. I'll add a pointer to the ssh key config below.

Note that the ignition config will only be processed on the very first boot - that's why we made a copy, so now we can restore our OS image from the pristine copy for successive experiments with CLC.

> As listed in the introduction above there are numerous options available for configuring Flatcar just the way you need it. For instance, you can specify a custom SSH key instead of your default one from `~/.ssh/` in the CLC config, by adding this section to your YAML file:

Reviewer: ```suggestion
As listed in the introduction above there are numerous options available for configuring Flatcar just the way you need it. For instance, you can specify a custom SSH key instead of your default one from ssh-agent/`~/.ssh/` in the CLC config, by adding this section to your YAML file:
```
Thanks, looks good now, I would just add both mechanisms that are used behind the scenes

        - ssh-rsa AAAAB......xyz email@host.net
```

Afterwards, transpile it again to Ignition JSON, overwrite `flatcar_production_qemu_image.img` with the fresh image file, and pass the ignition config to `./flatcar_production_qemu.sh` once again.

> ### Further reading: Platform / vendor specific information

Reviewer: I would add a short section here noting that auto updates are on by default and link to `setup/releases/update-strategies` for different update strategies.

Reviewer: maybe like

```
Flatcar has no package manager and performs its updates through an A/B partition mechanism.
The whole OS lives in `/usr` and updates to a new version are atomic because the whole OS gets updated to a new version this way.
It does auto updates by default and requires a reboot for them to apply.
The required reboot is done automatically through a reboot coordinator.
Other [update strategies][strategies] and are possible.

[strategies]: ../setup/releases/update-strategies
```

Author: Added.

[install-docs]: bare-metal/installing-to-disk
[ignition]: ../provisioning/ignition/
[cl-configs]: ../provisioning/cl-config
[cl-examples]: ../provisioning/cl-config/examples

> [cl-spec]: ../provisioning/config-transpiler/configuration

Reviewer: ```suggestion
[systemd]: ../setup/systemd/getting-started
```
